### PR TITLE
IA Committee Scraper

### PR DESCRIPTION
**Pull request by jonathanortega2023**
Implemented v1 of IA committee scraper, would probably benefit from more error handling and compartmentalizing.

### CODE REVIEW

File: scrapers_next/ia/committees.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                memberRole = member.rsplit(")", 1)[1].strip()
+                if len(memberRole) > 0:
+                    memberRole = memberRole.split(",", 1)[1].strip()
+                    com.add_member(name=memberName, role=memberRole)
+                else:
+                    com.add_member(name=memberName)
+            # Public members have names, role
+            elif "," in member:
+                memberName, memberRole = member.split(",", 1)
+                if len(memberRole) > 0:
```
Comment (johnseekins) [inline scrapers_next/ia/committees.py:34]: Values `> 0` are truth-y, as are non-empty strings. You _can_ say `if memberRole:` here and get the same effect.

File: scrapers_next/ia/committees.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                    com.add_member(name=memberName.strip(), role=memberRole.strip())
+            else:
+                memberName = member.strip()
+                if len(memberName) == 0:
```
Comment (johnseekins) [inline scrapers_next/ia/committees.py:38]: Similar to the note above, `0` and empty strings are false-y in python. So we could say `if not memberName:` here.

File: scrapers_next/ia/committees.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            classification = "subcommittee"
+            )
+        else:
+            com = ScrapeCommittee(
+            name=name,
+            chamber=chamber,
+            classification = "committee"
+            )
+        com.add_source(self.source.url)
+        # TODO Hard-coded
```
Comment (johnseekins) [inline scrapers_next/ia/committees.py:77]: If this is consistently where we're getting the committee listings from, I don't see why we can't leave it hard-coded.

File: scrapers_next/ia/committees.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        com.add_source("https://www.legis.iowa.gov/committees", "Committee list page")
+        # Get list of members contained in nested grid_12
+        # Translation hack in case they switch to all-caps
+        try:
+            membersList = XPath(
+                "//section/div/section[contains(@class, 'grid_12 alpha omega') and \
+                (contains(translate(., 'ABCDEFGHIJKLMNOPQRSTUVWXYZ','abcdefghijklmnopqrstuvwxyz'), 'house members') or \
+                contains(translate(., 'ABCDEFGHIJKLMNOPQRSTUVWXYZ','abcdefghijklmnopqrstuvwxyz'), 'senate members'))] //descendant::li"
+            ).match(self.root)
+        except:
```
Comment (johnseekins) [inline scrapers_next/ia/committees.py:88]: As the linting will tell you, python doesn't like "bare except"s. We need this to be _at least_ `except Exception:`.

**Review by NewAgeAirbender — APPROVED** [review]
Test scrape looks good, nice work!

**Comment by jonathanortega2023** [comment]
Is the failed linting black check a change I need to make?

File: scrapers_next/ia/committees.py [hunk 5]
```diff
@@ -1,97 +1,98 @@
...
+        com = CommitteeDetails(homeUrl, source=XPath("./a/@href").match_one(item))
+        return com
+
+
+class CommitteeDetails(HtmlPage):
+    def process_page(self):
+        # Helper method to parse member list item strings
+        def _parseMemberString(member: str, com: ScrapeCommittee):
+            if member.lower() == "no member data available":
+                return
```
Comment (johnseekins) [inline scrapers_next/ia/committees.py:21]: I think we may want to raise an EmptyScrape error or something here? (I can't remember the exact object for spatula...)

**Comment by chrisyamas** [comment]
> Hi @jonathanortega2023, could you copy/paste the below code block into your IA committees.py and add/commit/push that up?
> 
> Explanation: I had to make some modifications to your initial file for the data to work for our import process, but due to an issue (I believe with GitHub Codespaces) I was unable to make a pull request to your main branch of the forked repo (which would have needed to happen since you made this PR from your protected main branch).
> 
> Takeaway: Once you've pushed the below version of the committees scraper, we should be ready to approve and merge this PR! Thank you for all your hard work!!
> 
> ```
> from spatula import XPath, URL, HtmlListPage, HtmlPage, SkipItem, SelectorError
> from openstates.models import ScrapeCommittee
> import requests
> import lxml.html
> import re
> 
> 
> """
> The two scrapers that need to be run here are:
> - CommitteeList
> - JointAppropriationsComm
> 
> `CommitteeList` scrapes all committees and subcommittees, but
> `JointAppropriationsComm` scrapes the membership of each
> chamber's appropriations committee, and writes a committee object
> that is of classification="legislature" and membership consists of both
> upper and lower chamber appropriations committees members.
> --- There are two reasons for this
>     1. All scraped appropriations subcommittees are joint, and must correspond
>         to joint committee parent.
>     2. The appropriations committees of each chamber effectively coordinate
>         in a joint capacity, so this addition of additional comm reflects that.
> 
> IMPORTANT: when this jurisdiction's committee data are merged/imported into
> the database, the following steps should occur to avoid a KeyError when merging
> the joint appropriations subcommittees. The steps must occur  in this order:
>     1. merge only the one comm written from `JointAppropriationsComm`
>     2. move the JSON file for that comm into the folder with all the committees
>         scraped by `CommitteeList`
>     3. merge all committees
>     4. if a KeyError is raised, re-run each scraper,
>         and then repeat the above 3 steps
> """
> 
> 
> members_list_xpath = """
> //section/div/section[contains(@class, 'grid_12 alpha omega') and (contains
> (translate(., 'ABCDEFGHIJKLMNOPQRSTUVWXYZ','abcdefghijklmnopqrstuvwxyz'),
>  'house members') or contains(translate(., 'ABCDEFGHIJKLMNOPQRSTUVWXYZ',
>  'abcdefghijklmnopqrstuvwxyz'), 'senate members'))] //descendant::li
>  """
> 
> 
> def parse_member_string(member: str, com: ScrapeCommittee):
>     """
>     Helper function to parse member list item strings.
>     """
>     if member.lower() == "no member data available":
>         raise SkipItem("empty committee")
>     # Legislative members also have (party/district) info
>     if "(" in member:
>         member_name = member.split("(", 1)[0].strip()
>         member_role = member.rsplit(")", 1)[1].strip()
>         if member_role:
>             member_role = member_role.split(",", 1)[1].strip()
>             com.add_member(name=member_name, role=member_role)
>         else:
>             com.add_member(name=member_name)
>     # Public members simply have name, role
>     elif "," in member:
>         member_name, member_role = member.split(",", 1)
>         if member_role:
>             com.add_member(name=member_name.strip(), role=member_role.strip())
>     else:
>         member_name = member.strip()
>         if not member_name:
>             return
>         com.add_member(name=member_name)
> 
> 
> class UnknownChamberError(BaseException):
>     def __init__(self, com_name):
>         super().__init__(f"Chamber unknown for committee: {com_name}")
> 
> 
> class CommitteeDetails(HtmlPage):
>     def process_page(self):
>         # Page header holds name and chamber
>         header = (
>             XPath('//*[@id="content"]/div/section/h1')
>             .match_one(self.root)
>             .text.strip()
>             .rsplit(" ", 1)
>         )
>         name = header[0]
> 
>         if "s" in header[1].lower():
>             chamber = "upper"
>         elif "h" in header[1].lower():
>             chamber = "lower"
>         elif "j" in header[1].lower():
>             chamber = "legislature"
>         else:
>             raise UnknownChamberError(name)
> 
>         if "subcommittee" in name.lower():
>             classification = "subcommittee"
>         else:
>             classification = "committee"
> 
>         com = ScrapeCommittee(
>             name=name,
>             chamber=chamber,
>             parent="Appropriations" if classification == "subcommittee" else None,
>             classification=classification,
>         )
>         com.add_source(self.source.url)
>         com.add_source("https://www.legis.iowa.gov/committees", "Committee list page")
>         # Get list of members contained in nested grid_12
>         # Translation hack in case they switch to all-caps
>         try:
>             members_list = XPath(members_list_xpath).match(self.root)
>         except SelectorError:
>             raise SkipItem("No members found")
> 
>         for member in members_list:
>             member = member.text_content().strip()
>             parse_member_string(member, com)
> 
>         return com
> 
> 
> class CommitteeList(HtmlListPage):
>     source = URL("https://www.legis.iowa.gov/committees")
>     # Committee pages selector
>     selector = XPath('//*[@id="content"]/section/section/div/section/ul/li[*]')
> 
>     def process_item(self, item):
>         home_url = self.source.url
>         com = CommitteeDetails(
>             home_url, source=URL(XPath("./a/@href").match_one(item), timeout=30)
>         )
>         return com
> 
> 
> class JointAppropriationsComm(HtmlPage):
>     source = URL("https://www.legis.iowa.gov/committees")
>     name_and_role_re = re.compile(r"(.+)\s+\(.+\),\s+(.+)")
>     name_only_re = re.compile(r"(.+)\s+(\(.+\))")
> 
>     def process_page(self):
>         joint_com = ScrapeCommittee(
>             name="Appropriations", chamber="legislature", classification="committee"
>         )
> 
>         appr_coms = XPath(
>             ".//section[@class='divideVert grid_9 alpha omega'][1]//"
>             "li//a[text() = 'Appropriations']"
>         ).match(self.root)
> 
>         com_urls = [x.get("href") for x in appr_coms]
> 
>         for com_url in com_urls:
>             response = requests.get(com_url)
>             content = lxml.html.fromstring(response.content)
>             members_list = XPath(members_list_xpath).match(content)
> 
>             header = (
>                 XPath('//*[@id="content"]/div/section/h1')
>                 .match_one(content)
>                 .text.strip()
>                 .rsplit(" ", 1)
>             )
>             chamber_dict = {"H": "House", "S": "Senate"}
>             specific_chamber = chamber_dict[header[1][1]]
> 
>             joint_com.add_source(com_url, note=f"{specific_chamber} Committee Page")
> 
>             for member in members_list:
>                 member = member.text_content().strip()
>                 officer = self.name_and_role_re.search(member)
>                 if officer:
>                     name, partial_role = officer.groups()
>                     role = f"{partial_role}, {specific_chamber}"
>                 else:
>                     name_match = self.name_only_re.search(member).groups()[0]
>                     name, role = name_match, "member"
>                 joint_com.add_member(name=name, role=role)
> 
>         joint_com.add_source(self.source.url, "Committee list page")
>         joint_com.add_link(self.source.url, "homepage")
> 
>         return joint_com
> ```

In the interest of time, the required modifications were committed via [a separate pull request](https://github.com/openstates/openstates-scrapers/pull/4491) to `openstates-scrapers` repo. Consequently, this current PR is being closed without merge.

**Pull request closed without merging**